### PR TITLE
Merkle proof generation and validation

### PR DESCRIPTION
**Pull request by akorchyn**
Receipt tree Merkle proof.
The PR includes:
* Merkle proof creation
* Encoding of Branch, Extension, and Leaf nodes of Patricia merkle trie.

See also: https://github.com/ggxchain/cita-trie/pull/1

### CODE REVIEW

File: types/src/encode.rs [hunk 1]
```diff
@@ -17,3 +21,13 @@ macro_rules! encode {
         }
     };
 }
+
+/// Given an RLP encoded node, returns either RLP(node) or RLP(keccak(RLP(node)))
+pub fn rlp_node(rlp: &[u8], out: &mut dyn BufMut) {
+    println!("rlp_node: {:?}", rlp);
```
Comment (obsessed-cake) [inline types/src/encode.rs:27]: debug print?

File: types/src/receipt/trie/leaf.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#[derive(Debug, Clone)]
+pub struct ReceiptLeaf {
+    key: Vec<u8>,
+    value: TransactionReceipt,
+}
+
+impl ReceiptLeaf {
+    pub fn new(key: Nibbles, value: TransactionReceipt) -> Self {
+        dbg!(key.clone());
+        dbg!(alloy_rlp::encode(&value));
```
Comment (obsessed-cake) [inline types/src/receipt/trie/leaf.rs:14]: debug print?

File: types/src/encode.rs (hunk 1 above)
Comment (baidang201) [reply types/src/encode.rs:27]: may use log::info!

File: types/src/receipt/trie/leaf.rs (hunk 2 above)
Comment (baidang201) [reply types/src/receipt/trie/leaf.rs:14]: may use log::debug!

File: types/src/receipt/receipt_merkle_proof.rs [hunk 3]
```diff
@@ -56,44 +57,107 @@ pub enum ReceiptMerkleProofNode {
 /// from the leaf node.
 ///
 /// [1]: https://ethereum.org/se/developers/docs/data-structures-and-encoding/patricia-merkle-trie/
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
 pub struct ReceiptMerkleProof {
     pub proof: Vec<ReceiptMerkleProofNode>,
+    pub size: usize,
```
Comment (xcthulhu) [inline types/src/receipt/receipt_merkle_proof.rs:63]: Is this used anywhere?

File: types/src/receipt/transaction_receipt.rs [hunk 4]
```diff
@@ -1,16 +1,16 @@
...
+use serde::{Deserialize, Serialize};
 
-use crate::Bloom;
+use crate::{Bloom, Log};
 
-use super::log::Log;
 use super::tx_type::TxType;
 
 /// The receipt structure containing logs from smart contracts we are listening to; adapted from
 /// [`reth_primitives::ReceiptWithBloom`][1].
```
Comment (xcthulhu) [inline types/src/receipt/transaction_receipt.rs:9]:
```suggestion
/// [`reth_primitives::ReceiptWithBloom`][2].
```

My bad...

File: types/src/receipt/transaction_receipt.rs [hunk 5]
```diff
@@ -1,16 +1,16 @@
...
 ///
 /// [1]: https://ethereum.org/en/developers/docs/blocks/#block-anatomy
 /// [2]: https://github.com/paradigmxyz/reth/blob/f41386d28e89dd436feea872178452e5302314a5/crates/primitives/src/receipt.rs#L57-L62
```
Comment (xcthulhu) [inline types/src/receipt/transaction_receipt.rs:12]:
```suggestion
/// [1]: https://github.com/paradigmxyz/reth/blob/f41386d28e89dd436feea872178452e5302314a5/crates/primitives/src/receipt.rs#L57-L62
```

My bad

File: types/src/receipt/transaction_receipt.rs [hunk 6]
```diff
@@ -1,16 +1,16 @@
...
-#[derive(Debug, RlpEncodable, PartialEq, Clone)]
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
```
Comment (xcthulhu) [inline types/src/receipt/transaction_receipt.rs:13]: This will have to be `no_std` for the pallet, which will be a pain later...

Comment (akorchyn) [reply types/src/receipt/transaction_receipt.rs:13]: Smith did it, so I thought he tested it with a pallet

File: types/src/receipt/receipt_merkle_proof.rs (hunk 3 above)
Comment (akorchyn) [reply types/src/receipt/receipt_merkle_proof.rs:63]: Not anymore, good catch

File: types/src/receipt/transaction_receipt.rs (hunk 5 above)
Comment (akorchyn) [reply types/src/receipt/transaction_receipt.rs]: done

File: types/src/receipt/receipt_merkle_proof.rs (hunk 3 above)
Comment (akorchyn) [reply types/src/receipt/receipt_merkle_proof.rs]: done

File: types/src/receipt/trie/extension.rs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+    use cita_trie::MemoryDB;
+    use hasher::HasherKeccak;
+    use proptest::proptest;
+
+    use crate::{receipt::trie::leaf::ReceiptLeaf, Bloom, Log, Receipt, TransactionReceipt, H160};
+
+    use super::*;
+
+    proptest! {
```
[the remark this answers is not in the document]
Comment (akorchyn) [reply types/src/receipt/trie/extension.rs]: done

File: types/src/receipt/transaction_receipt.rs (hunk 6 above)
Comment (xcthulhu) [reply types/src/receipt/transaction_receipt.rs:12]: Sounds good

File: types/src/receipt/trie/extension.rs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+impl ExtensionNode {
+    pub fn new(prefix: Nibbles, pointer: H256) -> Self {
+        Self {
+            prefix: prefix.encode_path_leaf(false),
+            pointer,
+        }
+    }
+}
+
+impl ExtensionNode {
```
Comment (xcthulhu) [inline types/src/receipt/trie/extension.rs:23]:
```suggestion
```

**Review by xcthulhu — APPROVED** [review]
LGTM

**Pull request merged**